flush/sink.go

```diff
@@ -0,0 +1,152 @@
+// AckFunc returns an error, it will be returned by `Run`.
+func WithAckFunc(ackFn AckFunc) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackFn = ackFn
+	}
+}
+
+// WithMsgBufferSize sets the msg channel buffer size.
+func WithMsgBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush messages.
+type AsyncMessageSink struct {
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ams := &AsyncMessageSink{
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the context is done. If an AckFunc is configured, Run will execute it for each ack
+// recieved. If an error is returned, the user should cancel the context to prevent `Flush` from
+// blocking.
+func (ams *AsyncMessageSink) Run(ctx context.Context) error {
+	group, groupctx := rungroup.New(ctx)
+
+	group.Go(func() error {
+		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
+	})
+
+	group.Go(func() error {
+		for {
+			select {
+			case <-groupctx.Done():
+				return nil
+			case msg := <-ams.ackCh:
+				if ams.ackFn != nil {
+					err := ams.ackFn(msg)
+					if err != nil {
+						return err
+					}
+				}
+
+				atomic.AddUint64(&ams.acks, 1)
+			}
+		}
+	})
+
+	return group.Wait()
+}
+
+// PublishMessage publishes a message to the underlying sink. PublishMessage is desgined to be
+// called concurrently by the user.
+func (ams *AsyncMessageSink) PublishMessage(msg []byte) {
+	atomic.AddUint64(&ams.msgs, 1)
+	ams.msgCh <- message.NewMessage(msg)
```

Comment:
This should be done in a select, probably best with the context that's passed to the run function. Otherwise it might deadlock if the underlying sink dies...

It might be just my personal preferrence, but I would pass a context in when creating this wrapper and call the run function in a goroute there as well. This way the context can be stored in the wrapper and used here. Also the user has one less thing to worry about kinda like with the synchronous sink wrapper.

Reply:
I think it would be better of accepting a context rather than implicitly use only the one passed to the run function, there may be other factors that you might want the context cancelled for this higher up in the execution stack, like using the context from a GRPC request for instance. Maybe we could run this in a select on both the context from the `Run(...` and the user defined context?